cpu/cortexm_common/periph/systick.c

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright (C) 2020 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_periph_systick
+ * @{
+ *
+ * @file        systick.c
+ * @brief       Low-level timer driver implementation based on SysTick
+ *
+ *              SysTick is a 24-bit Down-Counting Timer
+ *              It runs at the same frequency as the CPU
+ *              The count (VAL) register can only be set to 0 on write
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ * @}
+ */
+
+#include <assert.h>
+#include "board.h"
+#include "cpu.h"
+#include "systick.h"
+
+/* SysTick CTRL configuration */
+#define ENABLE_MASK     (SysTick_CTRL_CLKSOURCE_Msk |\
+                         SysTick_CTRL_ENABLE_Msk    |\
+                         SysTick_CTRL_TICKINT_Msk)
+
+#define ALARM_EXT_MAX   ((1UL << (8 * sizeof(alarm_ext))) - 1)
+
+/* interrupt is always enabled to update time */
+static bool alarm_enabled;
+
+/* alarm callback */
+static systick_cb_t _cb;
+static void *_cb_arg;
+
+/* divider for CLOCK_CORECLOCK */
+static uint16_t prescaler;
+
+/* full periods before alarm */
+static uint16_t alarm_ext;
+
+/* remaining time to alarm (last period length) */
+static uint32_t alarm_val;
+
+/* current time if the timer ran at freq Hz */
+static uint32_t now;
+
+/* offset in CLOCK_CORECLOCK ticks */
+static uint32_t now_offset;
+
+void SysTick_timer_init(unsigned long freq, systick_cb_t cb, void *arg)
+{
+    assert(freq <= CLOCK_CORECLOCK);
+
+    _cb     = cb;
+    _cb_arg = arg;
+
+    prescaler = CLOCK_CORECLOCK / freq;
```

Comment:
Shouldn't you also assert that `freq` divides `CLOCK_CORECLOCK` without remainder? Otherwise frequency conversion is more involved. (And we have this implemented in properly `ztimer` and partially in `xtimer`. IMO we could just let SysTick run at CPU speed and leave frequency conversion to the higher level timer implementation.)

Reply:
I tried with `XTIMER_HZ` = 120 MHz first, but that lead to an "Unsupported hardware timer frequency" error.

Reply:
OK. But I think adding an `static_assert(CLOCK_CORECLOCK % freq == 0);` would be a good addition.

Reply:
Already added :)

Reply:
Hmm, somehow github shows me outdated diffs from time to time :-/ (Sometime it does show me a sneaky "refresh" button that solves the issue, but not every time.)